src/main/java/org/truffleruby/core/module/ModuleNodes.java

```diff
@@ -2335,15 +2334,23 @@ protected RubyArray usedRefinements() {
             final DeclarationContext declarationContext = RubyArguments.getDeclarationContext(frame);
             final Set<RubyModule> refinements = new HashSet<>();
             for (RubyModule[] refinementModules : declarationContext.getRefinements().values()) {
-                for (RubyModule refinementModule : refinementModules) {
-                    refinements.add(refinementModule);
-                }
+                Collections.addAll(refinements, refinementModules);
             }
             return createArray(refinements.toArray());
         }
 
     }
 
+    @CoreMethod(names = "refinements")
+    public abstract static class RefinementsNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        protected RubyArray refinements(RubyModule self) {
+            return createArray(self.fields.getRefinements().values().toArray());
```

Comment:
Needs a TruffleBoundary (on the specialization to keep it simple) for the `.values()` and `.toArray()`

Reply:
Added.